src/python/pants/backend/jvm/tasks/jvm_compile/javac/javac_compile.py

```diff
@@ -230,7 +232,7 @@ def _execute_hermetic_compile(self, cmd, ctx):
         exec_result.output_directory_digest,
         self.post_compile_extra_resources_digest(ctx, prepend_post_merge_relative_path=False),
       ])
-    classes_directory = ctx.classes_dir.path
+    classes_directory = Path(ctx.classes_dir.path).relative_to(get_buildroot())
```

Comment:
@wisechengyi or @stuhood, do either of you know what `CompileContext.classes_dir` should point to? In tests, it points to `{build_root}/.pants.d/compile/javac`, so this change of making the path be relative to the build root rather than an absolute path will work. But I want to make sure this doesn't break production.

Yi, for context, this PR is to require `DirectoryToMaterialize` to only use paths relative to the build root and then have the Rust engine prepend the build root path.

Reply:
`classes_directory` should always be under pants workdir, but pants workdir may not necessarily be under build root, so prepending buildroot to it in the engine may not work.

Reply:
> but pants workdir may not necessarily be under build root, so prepending buildroot to it in the engine may not work.

Oof. @gshuflin @cosmicexplorer @stuhood this is, then, a use case where we need to allow `DirectoryToMaterialize` to write to an arbitrary location, isn't it?

Reply:
> classes_directory should always be under pants workdir, but pants workdir may not necessarily be under build root, so prepending buildroot to it in the engine may not work.

This used to be the case, but at this point we are resigned to always having the workdir located below the buildroot. We should likely deprecate setting the workdir to anything that is not under the buildroot.

Reply:
> This used to be the case, but at this point we are resigned to always having the workdir located below the buildroot.

Why is this? 

> We should likely deprecate setting the workdir to anything that is not under the buildroot.

Hm, I support this, but we still couldn't land this PR until then, right? This PR will break Java compilation if the workdir is specified anywhere outside of the build root.

Reply:
It has been the case for a reasonably long time, because remote caching would break if you set it to something else.

Reply:
> It has been the case for a reasonably long time, because remote caching would break if you set it to something else.

Sounds good to me.

However, is it safe to assume that no one in the wild is using a special location and also compiling Java code? It feels wrong to me to merge this due to the breaking behavior, even though I would love to avoid a deprecation cycle. I don't know the JVM ecosystem very well, though, nor why we allowed a different workdir in the first place.

Reply:
Yes, that is safe to assume. We began moving away from anything else working about 2 years ago.

Reply:
Okay. I'll defer to your judgment here. Thank you for the review!